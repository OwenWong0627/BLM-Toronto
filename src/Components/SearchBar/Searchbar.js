import React from "react";

import usePlacesAutocomplete, {
   getGeocode,
   getLatLng,
} from "use-places-autocomplete";
import {
   Combobox,
   ComboboxInput,
   ComboboxPopover,
   ComboboxList,
   ComboboxOption,
} from "@reach/combobox";
import '@reach/combobox/styles.css';

//Search Bar
function Searchbar({ panTo, center, className }) {
   const {
      ready,
      value,
      suggestions: { status, data },
      setValue,
      clearSuggestions,
   } = usePlacesAutocomplete({
      requestOptions: {
         location: { lat: () => center.lat, lng: () => center.lng },
         radius: 200000,
      },
   });

   return (
      <div className={className}>
         <Combobox
            onSelect={async (address) => {
               //the setvalue keeps the text u inputted in the search box without having to call the api
               setValue(address, false);
               clearSuggestions();

               try {
                  const results = await getGeocode({ address });
                  const { lat, lng } = await getLatLng(results[0]);
                  panTo({ lat, lng });
                  console.log(lat, lng);
               }
               catch {
                  console.log("ERROR");
               }
            }}
         >
            <ComboboxInput
               value={value}
               onChange={(event) => {
                  setValue(event.target.value);
               }}
               disabled={!ready}
               placeholder="Enter an address"
            />
            <ComboboxPopover>
               <ComboboxList>
                  {status === "OK" &&
                     data.map(({ place_id, description }) => (
                        <ComboboxOption key={place_id} value={description} />
                     ))}
               </ComboboxList>
            </ComboboxPopover>
         </Combobox>
      </div>
   );
}

export default Searchbar;