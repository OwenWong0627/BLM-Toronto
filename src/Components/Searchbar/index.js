import React from "react";

import usePlacesAutocomplete, {
   getGeocode,
   getLatLng,
} from "use-places-autocomplete";
import {
   Combobox,
   ComboboxInput,
   ComboboxPopover,
   ComboboxList,
   ComboboxOption,
} from "@reach/combobox";
import '@reach/combobox/styles.css';
import PropTypes from 'prop-types';

/**
 * This component represents the searchbar in the Find a Business map page, and it takes in three props
 * The panTo prop is a function that, when clicked on a searched location, will pan to the searched location
 * The center prop allows the use-places-autocomplete API to give search results based on the user's location; it will give search results tailored to the user's location
 * The third prop name is just a class name for the css styling
 */
function Searchbar({ panTo, center, className }) {
   const {
      ready,
      value,
      suggestions: { status, data },
      setValue,
      clearSuggestions,
   } = usePlacesAutocomplete({
      requestOptions: {
         location: { lat: () => center.lat, lng: () => center.lng },
         radius: 200000,
      },
   });

   return (
      <div className={className}>
         <Combobox
            data-testid="search-bar"
            onSelect={async (address) => {
               //the setvalue keeps the text u inputted in the search box without having to call the api
               setValue(address, false);
               clearSuggestions();

               try {
                  const results = await getGeocode({ address });
                  const { lat, lng } = await getLatLng(results[0]);
                  panTo({ lat, lng });
                  console.log(lat, lng);
               }
               catch {
                  console.log("ERROR");
               }
            }}
         >
            <ComboboxInput
               data-testid="input"
               value={value}
               onChange={(event) => {
                  setValue(event.target.value);
               }}
               disabled={!ready}
               placeholder="Enter an address"
            />
            <ComboboxPopover>
               <ComboboxList data-testid="search-results">
                  {status === "OK" &&
                     data.map(({ place_id, description }) => (
                        <ComboboxOption key={place_id} value={description} />
                     ))}
               </ComboboxList>
            </ComboboxPopover>
         </Combobox>
      </div>
   );
}

/**
 * propType Documentation
 */
Searchbar.propTypes = {
   panTo: PropTypes.func.isRequired,
   center: PropTypes.object.isRequired,
   className: PropTypes.string.isRequired
}

export default Searchbar;
