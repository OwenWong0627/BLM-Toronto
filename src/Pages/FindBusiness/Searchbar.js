import React from "react";

import usePlacesAutocomplete, {
   getGeocode,
   getLatLng,
} from "use-places-autocomplete";
import {
   Combobox,
   ComboboxInput,
   ComboboxPopover,
   ComboboxList,
   ComboboxOption,
} from "@reach/combobox";
import '@reach/combobox/styles.css';

//Search Bar
function Searchbar({ panTo, center }) {
   //this is a hook
   //ready just checks whether the google scripts are all loaded up
   //value means what the user is currently inputting
   //suggestions means the suggestions that were returned from the google API
   //setValue is a function to set the value
   //clearSuggestion clears all the suggestions
   const {
      ready,
      value,
      suggestions: { status, data },
      setValue,
      clearSuggestions,
   } = usePlacesAutocomplete({
      requestOptions: {
         location: { lat: () => center.lat, lng: () => center.lng },
         radius: 200000,
      },
   });

   return (
      <div className="searchbar">
         <Combobox
            onSelect={async (address) => {
               //the setvalue keeps the text u inputted in the search box without having to call the api
               setValue(address, false);
               clearSuggestions();

               try {
                  const results = await getGeocode({ address });
                  const { lat, lng } = await getLatLng(results[0]);
                  panTo({ lat, lng });
                  console.log(lat, lng);
               }
               catch {
                  console.log("ERROR");
               }
            }}
         >
            <ComboboxInput
               value={value}
               onChange={(event) => {
                  setValue(event.target.value);
               }}
               disabled={!ready}
               placeholder="Enter an address"
            />
            <ComboboxPopover>
               <ComboboxList>
                  {status === "OK" &&
                     data.map(({ place_id, description }) => (
                        <ComboboxOption key={place_id} value={description} />
                     ))}
               </ComboboxList>
            </ComboboxPopover>
         </Combobox>
      </div>
   );
}

export default Searchbar;