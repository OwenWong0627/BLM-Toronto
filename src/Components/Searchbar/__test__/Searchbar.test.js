import React from 'react';
import ReactDOM from 'react-dom';
import Searchbar from '..';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import * as usePlacesAutocompleteModules from "use-places-autocomplete";

const mockCenter = { lat: 43.1, lng: -79.6 };

afterEach(cleanup);
it('Renders without crashing', () => {
   const div = document.createElement("div");
   ReactDOM.render((<Searchbar />), div);
});

it('Renders the searchbar and checks if the searchbar input area has a placeholder of Enter an address', () => {
   const { getByTestId } = render(<Searchbar />);
   expect(getByTestId("search-bar")).toHaveAttribute("placeholder", "Enter an address");
})

it('Renders the searchbar and checks if the manually set data values are displayed as text values in the searchbar combobox option', () => {
   usePlacesAutocompleteModules.default = jest.fn().mockReturnValue({
      ready: true,
      value: "",
      suggestions: { status: "OK", data: [{ place_id: "1", description: "Toronto, ON" }, { place_id: "2", description: "Mississauga, ON" }] },
      setValue: () => { },
      clearSuggestions: () => { }
   });
   const { getByText } = render(<Searchbar />);
   expect(getByText("Toronto, ON")).toBeInTheDocument();
   expect(getByText("Mississauga, ON")).toBeInTheDocument();
});

it('Renders the searchbar and checks if the manually set user input value is displayed in the search bar and has the correct suggestion', () => {
   usePlacesAutocompleteModules.default = jest.fn().mockReturnValue({
      ready: true,
      value: "Toron",
      suggestions: { status: "OK", data: [{ place_id: "1", description: "Toronto, ON, Canada" }, { place_id: "2", description: "Mississauga, ON, Canada" }] },
      setValue: () => { },
      clearSuggestions: () => { }
   });
   const { getByText, getByTestId } = render(<Searchbar />);
   expect(getByTestId("search-bar")).toHaveAttribute("value", "Toron");
   expect(getByText("Toron")).toBeInTheDocument();
   expect(getByText("to, ON, Canada")).toBeInTheDocument();
});