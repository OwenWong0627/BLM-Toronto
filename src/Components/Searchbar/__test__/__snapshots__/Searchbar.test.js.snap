// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    data-reach-combobox=""
  >
    <input
      aria-autocomplete="both"
      aria-controls="listbox--13"
      aria-expanded={false}
      aria-haspopup="listbox"
      data-reach-combobox-input=""
      data-testid="search-bar"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter an address"
      role="combobox"
      value="Toron"
    />
    <span />
  </div>
</div>
`;
